compiler/optimizer/VPHandlers.cpp

```diff
@@ -11252,6 +11249,30 @@ TR::Node *constrainSwitch(OMR::ValuePropagation *vp, TR::Node *node)
            node->setCannotOverflow(true);  // Let evaluator know value range of selector is covered by min and max case
          } // ... else if LongConstraint
       }
+   else
+      {
+      // TODO Can cases outside of the constraint range be removed?
+      TR::Node *selector = node->getFirstChild();
+      bool isGlobal;
+      TR::VPConstraint *constraint = vp->getConstraint(selector, isGlobal);
+      bool isInt64 = selector->getType().isInt64();
+      // TODO Is it worth it to handle 64 bit?
+      if (!isInt64 && constraint && constraint->asIntConstraint())
+         {
+         uint32_t numCases = node->getCaseIndexUpperBound() - 2;
+         int32_t low = constraint->asIntConstraint()->getLow();
+         int32_t high = constraint->asIntConstraint()->getHigh();
+         // Table entries are required to be contiguous. If numCases == constraint total range the selector is completely covered
+         if (numCases == ((high - low) + 1))
```

Comment:
This will incorrectly remove the bound check if the constraint doesn't line up exactly with the range of cases. As an extreme example, suppose we have 10 cases numbered [0, 9], and consider the constraint [100, 109]. We would remove the bound check even though the `table` is guaranteed to go to the default case. This is a problem even if there is overlap: take the same 10 cases and constraint [1, 10], and we'll remove the bound check even though a value of 10 will send control to the default case.

At the same time, there is a different way in which an equality test is too strict. Given the same 10 cases once again, and a constraint of [4, 7], we should know that the default case is dead (along with many of the other cases), but the equality test would fail to remove the bound check.

The proper condition to look for here is for the constraint to be a subset of the cases. If we have cases [a, b] and a (nonempty) constraint [c, d], then the switch is exhaustive iff a <= c and d <= b.

Reply:
yep I totally missed that in my re-write. I will update to the proper check!

Reply:
In all of my tests I was never saw a constraint that wasn't the range of an integer size. ie it was always -128 to 127, -32,768 to 32,767, etc so that is why I went with the cheapest check that would work if that held true. 

I guess from your explanations that a constraint could be for a much smaller range. hmmm for building my interpreter I wonder how I could modify my code so the switch selector gets a constraint of exactly my bytecode set.

Reply:
hmmm when table switches are created the case value is generally not stored since the selector is modified to create a 0 based index into the array. I am not sure I can get [a,b] values for the calculation....

Reply:
Judging by the (x86) evaluator for `table`, the cases are always contiguous starting at 0. So a is 0, and b is `numCases - 1`.

----

Probably the most common way to narrow the constraint on a value is to run branches based on inequality comparisons against the value, e.g.
```
if selector < 4 goto ...
if selector > 7 goto ...
use selector, which here is in [4, 7]
```
Of course, it won't help perf to introduce explicit branches into your interpreter for a better constraint, because the explicit branches will still be there. I think that telling the optimizer about the range of your bytecode values would require compiler changes, e.g. it would be possible to add an intrinsic that allows one to solemnly swear that a boolean (e.g. comparison result) is true, cross my heart, hope to invoke undefined behaviour

Reply:
The issue with using `a` as 0 since the table is 0 based is that most things appear to be signed types in JitBuilder (OMR compiler) so the constraint will always be from -MAX_<INT_TYPE> to +MAX_<INT_TYPE> and not 0 to MAX_UINT_TYPE. still poking around to see if I can come up with math that works.  Basically the only thing we can gleam from the table itself is the number of cases handled and that the cases have to be dense.  Maybe it is only possible to remove the bounds check if the constraint represents a full integer size range and the number of cases match. Everything else would require the actual case values.

Reply:
The integer ranges in VP all impose a signed interpretation, which means that certain ranges of unsigned integers (those crossing the max signed value for integers of the same size) are inexpressible using a single range. However, I doubt you have a jump table of 2^32 entries. I expect you're loading some narrower (8- or 16-bit) integer from memory and extending it to 32-bit. If so, then I think VP should create the constraint you want if you use zero-extension, e.g. [0, 255] in the case of zero-extending from 8-bits with `bu2i`

Reply:
VP does create the constraint at least in the case of `su2i`. Here's an example where the child of `su2i` was unconstrained:
```
   n17n su2i gets new constraint: value 17 is (0 to 65535)I
```
I expect it should work for `bu2i` as well since they use the same logic: [bu2i](https://github.com/eclipse/omr/blob/9e46da95f206f66a549b4f6df5b217f36fbc0bbc/compiler/optimizer/VPHandlers.cpp#L8674-L8678), [su2i](https://github.com/eclipse/omr/blob/9e46da95f206f66a549b4f6df5b217f36fbc0bbc/compiler/optimizer/VPHandlers.cpp#L8704-L8708)